### PR TITLE
Only specify `--target` by default for `-Zgcc-ld=lld` on wasm

### LINKED ISSUE #101653: -Z gcc-ld=lld fails again
I tried this code:

```console
RUSTFLAGS='-Z gcc-ld=lld' cargo +nightly build --release
```

I expected to see this happen: it builds my crate.

Instead, this happened: 

```console
 = note: rust-lld: error: library not found for -lcrt1.10.6.o
          clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Meta

`rustc --version --verbose`:
```console
rustc +nightly --version --verbose
rustc 1.65.0-nightly (1120c5e01 2022-09-08)
binary: rustc
commit-hash: 1120c5e01df508de64fe6642f22fadeb574afd6d
commit-date: 2022-09-08
host: x86_64-apple-darwin
release: 1.65.0-nightly
LLVM version: 15.0.0
```

I found in the invocation of `cc`

```console
nightly-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/bin/gcc-ld" "-fuse-ld=lld" "--target=x86_64-apple-macosx10.7.0"
```

```console
> clang --version
Apple clang version 13.1.6 (clang-1316.0.21.2.5)
Target: x86_64-apple-darwin21.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
> xcrun --sdk macosx --show-sdk-path
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.3.sdk
> cc --print-target-triple
x86_64-apple-darwin21.6.0
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

See https://github.com/rust-lang/rust/pull/101504#issuecomment-1242726753.

### PR DESCRIPTION
**Pull request by lqd**
On macOS, it's not yet clear which cases of clang/OS/target/SDK version impact how to find ld/lld/rust-lld. The `--target` argument is not needed on our current targets with a vanilla config to do so, but may be in some cases ? 

That is, things look to be different and more subtle than suggested in https://github.com/rust-lang/rust/issues/97402#issuecomment-1147404520.

Specifying this argument unconditionally currently breaks `-Zgcc-ld=lld` on the 10.7+ targets on x64 macOS. Vanilla configs on x64 and aarch64 don't seem to need it to be able to find `rust-lld`.

This fixes #101653 on macOS (I've tried on x64 and aarch64: vanilla installs with the most recent command line tools available for each of the OS versions I have access to, `run-make/issue-71519` passes).

I didn't expect the previous PRs to fail because of the existing tests, but CI does not actually run those tests yet, which explains the regressions. I was hoping to fix those in this PR but it's more involved (building lld is required for the tests to run, llvm/lld is not built on the test builders but on the dist builders, the dist builders don't run tests). This PR is just to unblock current users on macOS who have reported issues, but a regression could happen in the future by mistake until then.

r? @petrochenkov

### CODE REVIEW

**Comment by tschuett** [comment]
Maybe you will need the target on macOS for cross-compiling?

**Comment by lqd** [comment]
It wasn’t used before, so I’m not sure. And you can still pass it via -Clink-arg I presume ?

**Comment by Wyvern** [comment]
Looks like this problem still not fixed.

Another question. Why use `-fuse-ld=lld` generate `clang: error: invalid linker name in argument '-fuse-ld=lld` for now?

**Comment by lqd** [comment]
@Wyvern What do you mean not fixed, did you build this PR locally to test ? If so, what error did you get ?

**Comment by petrochenkov** [comment]
`10.7` comes from this function in `compiler\rustc_target\src\spec\apple_base.rs`:
```rust
fn macos_default_deployment_target(arch: &str) -> (u32, u32) {
    if arch == "arm64" { (11, 0) } else { (10, 7) }
}
```
, this is the default version that is used if no `MACOSX_DEPLOYMENT_TARGET` is provided by the user.

So setting `MACOSX_DEPLOYMENT_TARGET` that matches your environment should work as a quick workaround for https://github.com/rust-lang/rust/issues/101653.

**Comment by petrochenkov** [comment]
@Wyvern 
>Another question. Why use `-fuse-ld=lld` generate `clang: error: invalid linker name in argument '-fuse-ld=lld` for now?

This should only happen if your clang is old (or somehow modified) and doesn't support LLD. In that case it's reasonable that rustc using such clang doesn't support LLD either.

**Comment by petrochenkov** [comment]
So if no `MACOSX_DEPLOYMENT_TARGET` is specified, then rustc will always pass target `x86_64-apple-macosx10.7.0` to *LLVM* and apparently that's fine for all the range of supported versions.
However, it's apparently not fine to pass it to *clang* with a different default macosx version.

**Comment by petrochenkov** [comment]
>Maybe you will need the target on macOS for cross-compiling?

That's exactly why the `--target` option is passed, it's not necessary if we are not cross-compiling, clang should find the appropriate lld flavor for the host in that case.

There are several possible cases here:
- Not a cross compilation: `--target` is not necessary, the correct target is either hardcoded or used by default.
- Cross compilation:
    - Using a separate toolchain for the specific target (gcc is typically used like this), then `--target` is not necessary because it's hardcoded.
    - Using a toolchain that can switch targets dynamically (e.g. usual clang installed on the host), then `--target` is necessary for that clang to decide what triple to target.

**Comment by petrochenkov** [comment]
I think the initial fix should be
```rust
                    if we_are_cross_compiling() {
                        cmd.arg(format!("--target={}", sess.target.llvm_target));
                    }
```
, so the main question is how to implement the `we_are_cross_compiling()` part :D

`sess.host == sess.target` should work in most cases, but I'm not sure it will work for Apple targets specifically due to possible SDK version mismatches between the host and target.

**Comment by petrochenkov** [comment]
>so the main question is how to implement the `we_are_cross_compiling()`

`sess.target.lld_flavor != sess.host.lld_flavor`!

**Comment by tschuett** [comment]
This could miss the case where I am compiling from OSX to WatchOS. The lld flavour is the same, but I need the target parameter.

**Comment by petrochenkov** [comment]
@tschuett 
Sorry, I don't understand.
What do you do in this case when you are compiling without `-Zgcc-lld`? Exactly the same thing should work with `-Zgcc-lld` too.

**Comment by tschuett** [comment]
```console
>  cargo check --target x86_64-apple-darwin
```
I am on x86 OSX. The lld flavour is the same, but I would need the target parameter.

**Comment by Wyvern** [comment]
Is there a plan to merge this PR into nightly build?

**Comment by lqd** [comment]
@Wyvern Yes, we're working on it. Until then, you can also use the `MACOSX_DEPLOYMENT_TARGET` env var and it'll work.  Or a local lld via `-Clink-arg=-fuse-ld=lld` I guess.

**Comment by Wyvern** [comment]
`MACOSX_DEPLOYMENT_TARGET=12.0` works, Thx.

**Comment by lqd** [comment]
@petrochenkov I've rebased over #101988, and added the simili cross-compiling test for macOS, for the possibly different linker flavors between host and targets. Is this now what you expected ?

I've tested this PR on x64 darwin (without cross compilation shenanigans) and `-Zgcc-ld` will now work. I've also opened a draft PR, waiting on this one, to fix CI so that `needs-rust-lld` tests will be executed to prevent things from regressing again.

@rustbot ready

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -2831,11 +2831,31 @@ fn add_gcc_ld_path(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor) {
...
+                        // It could be required in some cases of cross-compiling with
+                        // `-Zgcc-ld=lld`, but this is generally unspecified, and we don't know
+                        // which specific versions of clang, macOS SDK, host and target OS
+                        // combinations impact us here.
+                        //
+                        // So we do a simple first-approximation until we know more of what the
+                        // Apple targets require (and which would be handled prior to hitting this
+                        // `-Zgcc-ld=lld` codepath anyway), but the expectation is that until then
+                        // this should be manually passed if needed.
+                        let is_targeting_wasm = matches!(flavor, LinkerFlavor::WasmLld(..));
```
Comment (petrochenkov) [inline compiler/rustc_codegen_ssa/src/back/link.rs:2853]: Not sure why WASM is special cased here, it is already covered by the linker flavor condition.

**Comment by petrochenkov** [comment]
Sorry for the delay, I'm currently taking some break from Rust.
@rustbot author

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 1 above)
Comment (lqd) [reply compiler/rustc_codegen_ssa/src/back/link.rs:2853]: It was less about special casing WASM than having to pick a bool between wasm and darwin, as we have to add `--target` on wasm, and not on darwin. 

So the choice was between `if wasm or different_linker_flavor` and `if !darwin or different_linker_flavor` and the first one seemed clearer to read but I can flip it to the second if you prefer ?

**Comment by lqd** [comment]
> Sorry for the delay, I'm currently taking some break from Rust.

@petrochenkov would you rather we re-roll for another reviewer ?

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_codegen_ssa/src/back/link.rs:2853]: I mean `different_linker_flavor` should always be true for WASM, so `wasm or different_linker_flavor == different_linker_flavor`.

**Comment by petrochenkov** [comment]
@lqd 
No, I reassigned most of my PRs so the remaining ones, including this, should be fine.

File: compiler/rustc_codegen_ssa/src/back/link.rs (hunk 1 above)
Comment (lqd) [reply compiler/rustc_codegen_ssa/src/back/link.rs]: Done

**Comment by lqd** [comment]
Removed the superfluous bool.

@rustbot ready

**Pull request merged**